HW2/ConsoleApp1/ConsoleApp1/Trie.cs

```diff
@@ -0,0 +1,156 @@
+namespace TrieDataStructure;
+
+/// <summary>
+/// Class of data structure Trie.
+/// </summary>>
+public class Trie
+{
+    private Node _root = new Node();
+    
+    private int size { get; set; }
+    
+    /// <summary>
+    /// Adds element to the trie.
+    /// </summary>>
+    /// <param name="element">Element to add.</param>
+    /// <returns>True if element hasn't been in trie before.</returns>
+    public bool Add(string element)
+    {
+        var node = this._root;
+        foreach (var letter in element)
+        {
+            if (!node.Children.ContainsKey(letter))
+            {
+                node.Children[letter] = new Node();
+            }
+            node = node.Children[letter];
+        }
+
+        if (node.IsTerminal)
+        {
+            return false;
+        }
+        node.IsTerminal = true;
+        this.size++;
+        return true;
+    }
+
+    /// <summary>
+    /// Checks is the element in trie.
+    /// </summary>>
+    /// <param name="element">Element to check.</param>
+    /// <returns>True if element in trie.</returns>
+    public bool Contains(string element)
+    {
+        var node = this._root;
+        foreach (var letter in element)
+        {
+            if (!node.Children.ContainsKey(letter))
+            {
+                return false;
+            }
+            node = node.Children[letter];
+        }
+        return node.IsTerminal;
+    }
+
+    /// <summary>
+    /// Removes element from the trie.
+    /// </summary>>
+    /// <param name="element">Element to remove.</param>
+    /// <returns>True if element really has been in trie.</returns>
+    public bool Remove(string element)
+    {
+        var (isElementInTree, isChildHasChildren) = DeleteRecursively(this._root, element, 0);
+        if (isElementInTree)
+        {
+            this.size--;
+        }
+        return isElementInTree;
+    }
+
+    private (bool, bool) DeleteRecursively(Node node, string element, int numberOfLetter)
+    {
+        if (node.IsTerminal && node.Children.Count == 0 && element.Length == numberOfLetter)
+        {
+            return (true, false);
+        }
+        if (node.IsTerminal && element.Length == numberOfLetter)
+        {
+            node.IsTerminal = false;
+            return (true, true);
+        } 
+        if (node.Children.Count == 0 && element.Length == numberOfLetter)
+        {
+            return (false, false);
+        }
```

Comment:
На мой взгляд так будет красивее:
```suggestion
        if (element.Lenght == numberOfLetter) {
            return (node.IsTerminal, node.Children.Count > 0);
        }
```
А куда здесь вставить `node.IsTerminal = false` -- это Вы придумайте.